Refactor AssetLoader and fix loader fbx and abc

kaamaurice opened the pull request:
## Brief description
Refactor AssetLoader and fix loader for fbx and abc representation.

kaamaurice committed:
Blender api update for workflow using override lib

kaamaurice committed:
Blender plugins creators update

kaamaurice committed:
Blender plugins loaders update

kaamaurice committed:
Blender plugins publish collectors and increment

kaamaurice committed:
Blender plugins publish validators for new workflow with actions

kaamaurice committed:
Openpype pipeline constant AVALON_INSTANCE_ID added

kaamaurice committed:
Openpype plugins publish integrate_new can get representation id from an earlier stage

kaamaurice committed:
Blender default settings and schemas

kaamaurice committed:
Blender default settings and schemas

kaamaurice committed:
Blender artist docs update

kaamaurice committed:
Blender plugins publish extractors update

kaamaurice committed:
Minor refactor and notes

kaamaurice committed:
remove register_target local since it was implemented in context_tools

kaamaurice committed:
refactor Blender load setdress with simple plugin.AssetLoader

kaamaurice committed:
Update maintained parent and modifiers

kaamaurice committed:
TODO comment for maintained modifiers order

kaamaurice committed:
maintained modifiers order done

kaamaurice committed:
refactor maintained contextmanager in class

kaamaurice committed:
bugfix load_animation update_layout and namespace issues

kaamaurice committed:
bugfix Blender startup register def and quit with base module error

kaamaurice committed:
restor set_start_end_frames function

kaamaurice committed:
bugfix SelectInvalidAction

kaamaurice committed:
bugfix Extract_layout with no animation

kaamaurice committed:
fix aUpdateContainer using new inventory and refactor ContainerMaintainer

kaamaurice committed:
fixbug with is_system_override and clean pprint

kaamaurice committed:
link and append loader for blend

kaamaurice committed:
remove Creator use_selection armature to collection

kaamaurice committed:
Finalized append link Loader with icon and order

kaamaurice committed:
Update project settings Blender Publish Validator and Extractor

kaamaurice committed:
Update rig loaders and assets validators

kaamaurice committed:
updated animation loader extractor and project settings schemas

kaamaurice committed:
blend workfile can be linked or appended

kaamaurice committed:
Save workfile action

kaamaurice committed:
implemented switch blend assets

kaamaurice committed:
fix metadata asset_name

kaamaurice committed:
fix switch for instance collection

kaamaurice committed:
fix update_metadata

kaamaurice committed:
load layout with animation instances link mode settings

kaamaurice committed:
Blender Creators defaults subset list

kaamaurice committed:
Remove MODEL_DOWNSTREAM constant

kaamaurice committed:
new inventory Set Version with local data transfert

kaamaurice committed:
note update for inventory transfert data

kaamaurice committed:
revert bugfix from module base

kaamaurice committed:
extract blend pack unpack fixed

kaamaurice committed:
fix duplicate blender settings keys

kaamaurice committed:
fix bug merge conflict with last rebase

kaamaurice committed:
fix blender settings and schemas

kaamaurice committed:
disable setdress integrate animation

kaamaurice committed:
switching asset update obj name and namespace

kaamaurice committed:
refactor context override with temp_override

Tilix4 committed:
Refactor: _update_process consistency with _process

Tilix4 committed:
correct typing

Tilix4 committed:
change audio datablock removal and removed libpath from exec_update return

Tilix4 committed:
Apply suggestions from code review

Co-authored-by: kaa <kaalium@gmail.com>

kaamaurice committed:
fix switch rename

Tilix4 committed:
Feature: Blender Seamless integration

Tilix4 committed:
Add load handler to update loader metadata

kaamaurice committed:
neasted plublishable instances

kaamaurice committed:
use file compress on save with OP settings

kaamaurice committed:
use file compress on save with OP settings

kaamaurice committed:
clean print

Tilix4 committed:
Moved loader matching to handler

Tilix4 committed:
Fix: Embedded libraries don't remap correctly

Tilix4 committed:
Fix: Embedded linked libraries are not correctly remapped between remote and studio

Tilix4 committed:
Docstring

Tilix4 committed:
Add blender properties

Tilix4 committed:
Revert namespace naming change

kaamaurice committed:
fix make_paths_relative script

Tilix4 committed:
Base implementation of native UI creator

Tilix4 committed:
Variant name with defaults

Tilix4 committed:
Asset name restricted to enum

kaamaurice committed:
fix error messagebox stay on top

kaamaurice committed:
fix workfile breaked relative path

Tilix4 committed:
create openpype instance and handle outliner for relevant data

Tilix4 committed:
asset name prop search

Tilix4 committed:
append datablocks to existing instance

kaamaurice committed:
clean code

kaamaurice committed:
fix make paths relative

Tilix4 committed:
Feature: Pre-Launch convert blend file paths to absolute before opening

Tilix4 committed:
Use setting for both publish and load

Tilix4 committed:
Documentation

Tilix4 committed:
fix

Tilix4 committed:
Fix: Sync copy made before paths are made absolute

Tilix4 committed:
Feature: Operator to make a collection container publishable (#21)

* Feature: Operator to make a collection container publishable

* correct operator name

* added to collection and outliner menus and UI selection removed

Tilix4 committed:
Path as posix and moved to blender hooks

Tilix4 committed:
Fix: Make container publishable None found

Tilix4 committed:
clean imports

Tilix4 committed:
consistent getting of collection

kaamaurice committed:
clean code

Tilix4 committed:
UIList in Scene space context

Tilix4 committed:
Scene UI to manage instances

Tilix4 committed:
Collect instances

Tilix4 committed:
fix extract animation

Tilix4 committed:
fix collect instance to publish

kaamaurice committed:
fix: hook mod vertex_indices_set exlude on update

Tilix4 committed:
Renaming datablocks with instance suffix

Tilix4 committed:
update representation by passing arguments

kaamaurice committed:
fix extract blend pack image with generated img

Tilix4 committed:
Fix: relative paths conformation

Tilix4 committed:
comments and black

Tilix4 committed:
remap images

Tilix4 committed:
ensure loop

Tilix4 committed:
Add OpenpypeContainer propertygroup

Tilix4 committed:
fix get_loader_name

kaamaurice committed:
added settings for validator no colons in name

Tilix4 committed:
Fix: instances embedded in container are published

Tilix4 committed:
code style

Tilix4 committed:
fix containers update not run if not path management

Tilix4 committed:
don't containerize hierarchy for outliner

Tilix4 committed:
ls() ensures duplicated/existing containers

Tilix4 committed:
Cleaning

Tilix4 committed:
Props description

Tilix4 committed:
Fix: Published workfile relative paths are not correctly remapped

Tilix4 committed:
clean containers list update

Tilix4 committed:
refactor load process design

Tilix4 committed:
default load_type to None and raises error if not implemented

Tilix4 committed:
clean create process

Tilix4 committed:
update refactor

Tilix4 committed:
use pointer property to library

Tilix4 committed:
switch() refactor

Tilix4 committed:
Container setter and getter for name

Tilix4 committed:
fix containers auto creation

Tilix4 committed:
refactor collections names for datapath

Tilix4 committed:
fix replace container behaviour

Tilix4 committed:
fix getter

Tilix4 committed:
unlink_from_collection util function

Tilix4 committed:
keep collection parenting

Tilix4 committed:
fix update representations with getattr)

Tilix4 committed:
purge instances without datablocks at save

Tilix4 committed:
purge instances without datablocks at save

Tilix4 committed:
instances creation with new datablocks referencement

Tilix4 committed:
use Path

Tilix4 committed:
put fake user to datablocks referenced in instance

Tilix4 committed:
renaming to datablock_refs

Tilix4 committed:
ft publish bool to disable instance from publish

Tilix4 committed:
fix out of range for active index

Tilix4 committed:
fix fake user on deleted datablock

Tilix4 committed:
optimize creators listing

Tilix4 committed:
refactor creator, collect and extract for outliner data

Tilix4 committed:
change collections colors for only two

Tilix4 committed:
remove fake user from loaded datablocks

Tilix4 committed:
outliner workflow

Tilix4 committed:
rig workflow

Tilix4 committed:
fix published colors

Tilix4 committed:
fix no datablock refs in ui

Tilix4 committed:
layout loading refactor and animation instance handles objects

Tilix4 committed:
fix rig new action name

Tilix4 committed:
look loader imported functions naming

Tilix4 committed:
fix replace container optimization relink with datablock_refs

Tilix4 committed:
animation workflow

Tilix4 committed:
fix camera creation use selection empty

Tilix4 committed:
remove exec_process useless decorator

Tilix4 committed:
refactor workfile loader. WIP for APPEND

Tilix4 committed:
WIP for non-blender loaders

Tilix4 committed:
fix error no scene attribute at integration

Tilix4 committed:
cleaning

Tilix4 committed:
remove animation instance creation for each armature from load blend layout

Tilix4 committed:
fix container datablocks publish

Tilix4 committed:
ui tweaks: max rows to 5 and checkbox before name

Tilix4 committed:
ui tweaks: add/remove moved to right side

Tilix4 committed:
ui tweaks: icon before datablock name

Tilix4 committed:
instance checkbox disables in pyblish instead of removing

Tilix4 committed:
fix container created from collections not in scene

Tilix4 committed:
fix missing creator name

Tilix4 committed:
comment

Tilix4 committed:
ui check validity of datablock type

Tilix4 committed:
move datablocks uilist to subpanel

Tilix4 committed:
ui bl types icons moved to bottom

Tilix4 committed:
ui subset name drawn and full name below

Tilix4 committed:
ui : move up/down buttons for instances and datablocks

Tilix4 committed:
refactor extract workfile to move images to resources

Tilix4 committed:
hard disable pack

Tilix4 committed:
fix switch errors and better datablocks remapping

Tilix4 committed:
loaders type updated before container auto creation

Tilix4 committed:
fix unsaved filepath error

Tilix4 committed:
loader auto assigned for auto created containers

Tilix4 committed:
remove useless children getting for outliner instances

Tilix4 committed:
update hero version

Tilix4 committed:
simplify integrate blend

Tilix4 committed:
fix datablock user remap

Tilix4 committed:
fix materials messing after load

Tilix4 committed:
fix old and new datablock matching

Tilix4 committed:
process images files not packed

Tilix4 committed:
fix hero version no _id

Tilix4 committed:
fix IDs referenced by main datablocks

Tilix4 committed:
fix hero version missing blender. families

Tilix4 committed:
CollectResourcesPath.families uses IntegrateAsset.families

Tilix4 committed:
fix published resources with wrong paths

Tilix4 committed:
fix None for datapath IntegrateBlend

Tilix4 committed:
fix make abs paths

Tilix4 committed:
fix recursive orphan purge messes the file

Tilix4 committed:
fix import entities

Tilix4 committed:
fix ui error message no openpype instance

Tilix4 committed:
add datablock to instance warn user if datablock already in instance

Tilix4 committed:
Feature: Prefill datacol with active object or collection

Tilix4 committed:
Feature: Gather into collection checkbox for instances datablocks

kaamaurice committed:
Fix loader fbx and abc

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -1191,24 +1191,111 @@ def _load_library_datablocks(
 
         container_collection = bpy.data.collections.get(container_name)
+
+        if not container_collection:
```
Ne devrait-on pas conditionner la colloction à plusieurs objets ? Si un seul objet alors l'importer tel quel.

kaamaurice replied:
Si on se libère de la collection pour définir un container OpenPype alors oui on peux imaginer importer sans collection un objet, ou même plusieurs objet, mais si on veux avoir deux comportement diffèrent alors optons pout deux loaders : `import as collection` / `import as objects set`

Tilix4 reviewed (commented):
Rien à dire sauf une petite question qui m'est venue à l'esprit !

kaamaurice committed:
Fix seamless with no namespace and no maintainer (#43)

* Fix seamless namespace and maintainer

* fix reviewed code and clean last maintainer params

kaamaurice committed:
fix fbx update and switch

Tilix4 committed:
Refactor: Don't create a default collection at loading (#45)

kaamaurice committed:
fix rename when update

kaamaurice committed:
fix abc loader and try to deal with occurences

Tilix4 committed:
Fix: Workfile have hero versions published (#49)

Tilix4 committed:
Fix: Camera instance creation with gather into collection (#48)

Tilix4 committed:
Fix: Gather into collection available when Collection is allowed (#46)

* Fix: Gather into collection available when Collection is allowed

* fix context argument

Tilix4 committed:
Refactor: make_container_publishable. (#47)

- Created get_all_outliner_children to ease children/objects listing
- Moved presetting of create_openpype_instance from init to invoke for reliability

Tilix4 committed:
Refactor: Created _containerize_datablocks function to ease _load_library_datablocks overrides (#51)

Tilix4 committed:
Fix: camera create by matching camera from datablocks and all children (#53)

* Fix: camera create by matching camera from datablocks and all children

* refactor with _get_camera_from_datablocks as shared function

Tilix4 committed:
Feature: make_container_publishable can convert container to current asset. (#54)

It keeps the subset name but changes the asset to the current one. This is useful to 'make local' OP containers and use them as instances for the current asset.
- Also fixes the way collections are nested or converted
- Added a missing docstring

Tilix4 committed:
Change: Remove action auto assignation to objects (#56)

* Change: Remove action auto assignation to objects

* fix loader listing

Tilix4 committed:
Fix: Gather into collection checkbox display (#58)

kaamaurice committed:
fix: added extract anim abc in project settings

Tilix4 committed:
Fix: ExtractBlendAnimation no animation_data (#60)

Tilix4 committed:
Feature: Expose container content

Commit:
Merge branch 'ft_expose_container_content' into 'enhancement/OP_Blender-dynamic-workflow-using-linked-libraries'

Feature: Expose container content

See merge request kaamaurice/open-pype-normaal!1

Sharkitty committed:
Enhancement: Set object mode on extract blend

Tilix4 committed:
Feature: Settings for Setdress creator and loader

Tilix4 committed:
Merge branch 'ft_setdress_settings' into 'enhancement/OP_Blender-dynamic-workflow-using-linked-libraries'

Feature: Settings for Setdress creator and loader

See merge request kaamaurice/open-pype-normaal!2

Tilix4 committed:
Merge branch 'enhancement_extract_blend_set_object_mode' into 'enhancement/OP_Blender-dynamic-workflow-using-linked-libraries'

Enhancement: Set object mode on extract blend

See merge request kaamaurice/open-pype-normaal!3

Tilix4 committed:
Fix: duplicate ValidateNoColonsInName

Tilix4 committed:
Merge branch 'fx_duplicate_validator' into 'enhancement/OP_Blender-dynamic-workflow-using-linked-libraries'

Fix: duplicate ValidateNoColonsInName

See merge request kaamaurice/open-pype-normaal!4

Tilix4 committed:
Fix: Wrong context object.mode_set (#62)

Tilix4 committed:
Fix: action not user remapped (#65)

* Fix: action not user remapped

* rewrite comment because not a bug

* black

Tilix4 committed:
Fix: Expose Container Content label (#63)

Tilix4 committed:
Fix: published images resources uncorrectly mapped (#55)

kaamaurice committed:
refactor 1st part

Tilix4 committed:
Fix: Container representation ID overridden by update_representations (#64)

* Fix: Container representation ID overriden by update_representations

* add exception for datablock with representation assigned

Tilix4 committed:
Feature: make_container_publishable support linked outliner_entity (#66)

Now uses expose_container_content as base

Tilix4 committed:
Fix: get_all_outliner_children to accept only outliner types (#68)

* Fix: get_all_outliner_children to accept only outliner types

* TypeError

kaamaurice committed:
refactor AssetLoader and loader abc fbx

kaamaurice committed:
clean rebase conflicts

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -636,14 +638,7 @@ class Loader(LoaderPlugin):
-    The `update` method should be implemented by a sub-class, because
-    it's different for different types (e.g. model, rig, animation,
-    etc.).
+    """A basic AssetLoader for Blender.
```
In this case I think it must be called `BlenderLoader` if dedicated to load from Blender files.

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -689,116 +684,33 @@ def _get_scene_container(
+    def _containerize_objects_in_collection(
         self,
-        libpath: Path,
         container_name: str,
```
`container_name` given value is never used but always replaced by `outliner_entity.name`. It is the expected behaviour IMO, therefore remove this argument.

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -689,116 +684,33 @@ def _get_scene_container(
-                ),
-                None,
-            )
+        objects: List[bpy.types.Object],
```
It is not only objects but could be `collections` too. Stick to `outliner_entities` naming.

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -689,116 +684,33 @@ def _get_scene_container(
-                container_name = outliner_entity.name
-        else:
-            outliner_entity = None
+        datablocks = set(objects)
```
`outliner_entities` must be a set when provided to the function and not being cast.

kaamaurice committed:
fix loader with blend libraries

kaamaurice committed:
fix extract abc and fbx plugins

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -628,6 +629,7 @@ def _remove_instance(self, instance_name: str) -> bool:
         return True
 
+
 class Loader(LoaderPlugin):
```
This class is useless, I think `AssetLoader` (renamed `BlenderLoader`) could be the one inheriting from `LoaderPlugin`

Tilix4 commented on openpype/hosts/blender/api/plugin.py:
```diff
@@ -1345,6 +1039,358 @@ def remove(self, container: Dict) -> bool:
+
+        return container, all_datablocks
+
+    def _instance_blend(
```
Because we are refactoring, this one could be moved right after the `_link_blend`.

Tilix4 reviewed (changes requested):
Nice refactor! Few changes still, but we are getting very close to something polished I'd say.
